backend/server.go

```diff
@@ -182,63 +192,58 @@ var roleCast = map[string]gamelogic.Role{
 	"doppleganger": gamelogic.DoppleGanger,
 }
 
-func AssignRolePool(game *gamelogic.Game, responseWriter http.ResponseWriter, request *http.Request) (error, int) {
+func AssignRolePool(game *gamelogic.Game, responseWriter http.ResponseWriter, request *http.Request) (JsonBody, HttpStatus, error) {
 	rolesReq := struct{ Roles []string }{}
 
 	if err := json.NewDecoder(request.Body).Decode(&rolesReq); err != nil {
-		return InvalidJSON, http.StatusBadRequest
+		return nil, http.StatusBadRequest, InvalidJSON
 	}
 
 	var roles []gamelogic.Role
 
 	for _, roleName := range rolesReq.Roles {
 		role, ok := gamelogic.RoleNameToID[roleName]
 		if !ok {
-			return RoleNoteFoundError, http.StatusBadRequest
+			return nil, http.StatusBadRequest, RoleNoteFoundError
 		}
 		roles = append(roles, role)
 	}
 
 	game.AssignRolePool(roles)
-	fmt.Fprint(responseWriter, "{}")
 
-	return nil, http.StatusOK
+	return emptyJson, http.StatusOK, nil
 }
 
-func StartGame(game *gamelogic.Game, responseWriter http.ResponseWriter, request *http.Request) (error, int) {
+func StartGame(game *gamelogic.Game, responseWriter http.ResponseWriter, request *http.Request) (JsonBody, HttpStatus, error) {
 
 	if err := game.Start(); err != nil {
-		return err, http.StatusBadRequest
+		return nil, http.StatusBadRequest, err
 	}
 
 	go game.ExecuteNight()
 
-	fmt.Fprint(responseWriter, "{}")
-	return nil, http.StatusOK
+	return emptyJson, http.StatusOK, nil
 }
 
-func CreateGame(responseWriter http.ResponseWriter, request *http.Request) (error, int) {
+func CreateGame(responseWriter http.ResponseWriter, request *http.Request) (JsonBody, HttpStatus, error) {
 	name := "Game1"
 
 	game, err := gamelogic.CreateGame(name)
 	if err != nil {
-		return err, http.StatusInternalServerError
+		return nil, http.StatusInternalServerError, err
 	}
 
 	games.Store(name, game)
 
-	json.NewEncoder(responseWriter).Encode(map[string]string{
+	body := map[string]interface{}{
 		"id": game.Id,
-	})
+	}
 
-	return nil, http.StatusOK
+	return body, http.StatusOK, nil
 }
 
-func GetPlayerInfo(player *gamelogic.Player, game *gamelogic.Game, responseWriter http.ResponseWriter, request *http.Request) (error, int) {
-	if err := json.NewEncoder(responseWriter).Encode(player); err != nil {
-		return err, http.StatusInternalServerError
-	}
-	return nil, http.StatusOK
+func GetPlayerInfo(player *gamelogic.Player, game *gamelogic.Game, responseWriter http.ResponseWriter, request *http.Request) (JsonBody, HttpStatus, error) {
+	return player.ToJson(), http.StatusOK, nil
```

Comment:
Instead of passing player to json. Why not pass a map[name]=player object. That way we can keep the JSON Marshall function in the player object.

Reply:
i'm not sure i understand the suggestion. Are you suggesting doing the following?
```go
const playerJson = map[string]interface{}{
  "player": player,
}
return playerJson, htpt.StatusOK, nil
```

I'm very weakly opposed to this because it has the effect of nesting the all the info a layer deeper. I.e. the json ends up looking like this
```
{
  "player": {
    "actionPrompt": "choose-center-card"
    "hasSeen": {}
    "name": "player 1"
    "originalRole": "werewolf"
  }
}
```

instead of like this
```
{
  "actionPrompt": "choose-center-card"
  "hasSeen": {}
  "name": "player 1"
  "originalRole": "werewolf"
}
```

Also, i feel like the JSON Marshall function is still in the player object. I've just renamed it to ToJson.

I don't really care about this one way or the other though. If you have a strongish opinion about having JSONMarshal in in player and using the nested json, i'm fine with that

Reply:
If i remember correctly having a JSONMarshal function fulfills the JSON serializable interface. So i'd like to keep that method. I see your point that nesting player is kinda gross. I just like the idea that Wrap API ENdpoint deals with objects -> JSON. So theoretically if we decided to move to XML or something else only the WrapAPIEndpoint would have to change.

Reply:
ah here's an idea (actually had the idea earlier but I couldn't find the interface i was looking for).

We can make ApiHandlerFunc return type `json.Marshaler`
```go
type ApiHandlerFunc func(http.ResponseWriter, *http.Request) (json.Marshaller, HttpStatus, error)
````
https://golang.org/pkg/encoding/json/#Marshaler

that way GetPlayerInfo can return the player directly (keeping the flat structure). And other things can return their own custom `map[string]interface{}`'s. Since `map[string]interface{}` imlements `json.Marshaler`

Reply:
Apparently `map[string]interface{}` doesn't implement `json.Marshaler`. Which is really confusing. I don't understand why I can't do this
```go
var foo json.Marshaler
foo = map[string]interface{}{}
```

but I can do this
```go
json.Marshal(map[string]interface{}{})
```

i've spent way longer on this than i should have